source/crud/bulk-write.md

```diff
@@ -0,0 +1,724 @@
+     * occurring on individual writes.
+     */
+    error: Optional<Error>;
+
+    /**
+     * Write concern errors that occurred while executing the bulk write. This list may have
+     * multiple items if more than one server command was required to execute the bulk write.
+     */
+    writeConcernErrors: WriteConcernError[];
+
+    /**
+     * Errors that occurred during the execution of individual write operations. This map will
+     * contain at most one entry if the bulk write was ordered.
+     */
+    writeErrors: Map<Int64, WriteError>;
+
+    /**
+     * The results of any successful operations that were performed before the error was
+     * encountered.
+     */
+    partialResult: Optional<BulkWriteResult>;
+}
+```
+
+## Building a `bulkWrite` Command
+
+The `bulkWrite` server command has the following format:
+
+```json
+{
+    "bulkWrite": 1,
+    "ops": <Array>,
+    "nsInfo": <Array>,
+    "errorsOnly": <Boolean>,
+    "ordered": <Boolean>,
+    "bypassDocumentValidation": <Boolean>,
+    "comment": <Bson>,
+    "let": <Document>,
+    ...additional operation-agnostic fields
+}
+```
+
+Drivers SHOULD use document sequences ([`OP_MSG`](../message/OP_MSG.rst) payload type 1) for the
+`ops` and `nsInfo` fields. Drivers MUST NOT use document sequences when auto-encryption is enabled.
+
+The `bulkWrite` command is executed on the "admin" database.
+
+### Operations
+
+The `ops` field is a list of write operation documents. The first entry in each document has
+the name of the operation (i.e. "insert", "update", or "delete") as its key and the index in the
+`nsInfo` array of the namespace on which the operation should be performed as its value. The
+documents have the following format:
+
+#### Insert
+
+```json
+{
+    "insert": <Int32>,
+    "document": <Document>
+}
+```
+
+Drivers MUST add an `_id` field at the beginning of the insert document if one is not already
+present.
+
+When a user executes a bulk write with an unacknowledged write concern, drivers SHOULD check the
+size of the insert document to verify that it does not exceed `maxBsonObjectSize`. For acknowledged
+bulk writes, drivers MAY rely on the server to return an error if the document exceeds
+`maxBsonObjectSize`. The value for `maxBsonObjectSize` can be retrieved from the selected server's
+`hello` response.
+
+#### Update
+
+```json
+{
+    "update": <Int32>,
+    "filter": <Document>,
+    "updateMods": <Document | Array>,
+    "multi": <Boolean>,
+    "upsert": Optional<Boolean>,
+    "arrayFilters": Optional<Array>,
+    "hint": Optional<String>
+}
+```
+
+#### Delete
+
+```json
+{
+    "delete": <Int32>,
+    "filter": <Document>,
+    "multi": <boolean>,
+    "hint": Optional<String>,
+    "collation": Optional<Document>
+}
+```
+
+### Namespace Information
+
+The `nsInfo` field is an array containing the namespaces on which the write operations should be
+performed. Drivers MUST NOT include duplicate namespaces in this list. The documents in the
```

Comment:
Is the prohibition of duplicate namespaces enforced, or are we on the honor system here?